### PR TITLE
No need to redirect to the original frame explicitly during deserialization

### LINKED ISSUE #3721: Improve Universe serialization performance
## Is your feature request related to a problem? ##
The speed and memory footprint of `Universe` serialization greatly influences the performance of parallel analysis especially when the analysis block is small. This Issue serves as an (ongoing) collection of possible improvements for the performance.

## Benchmark ##

```python
 # 560k atom, 125880 residues
u = mda.Universe('test.pdb')

dump_times = np.zeros(10)
load_times = np.zeros(10)

for rep in range(10):
    start_time = time.time()

    u_dumped = pickle.dumps(u)
    dump_finish_time = time.time()
    dump_times[rep] = dump_finish_time - start_time
    
    u_new = pickle.loads(u_dumped)
    load_finish_time = time.time()
    load_times[rep] = load_finish_time - dump_finish_time

print(f"dump time: {dump_times.mean():.2f} \u00B1 {dump_times.std():.2f} s")
print(f"load time: {load_times.mean():.2f} \u00B1 {load_times.std():.2f} s")

> dump time: 0.45 ± 0.00 s
> load time: 1.48 ± 0.06 s

> dump only trajectory: 0.00 ± 0.00 s
> load only trajectory: 1.14 ± 0.07 s

> dump only topology: 0.46 ± 0.03 s
> load only topology: 0.19 ± 0.01 s
```

```python
with open('universe.pkl', 'wb') as f:
    pickle.dump(u, f)
!du -sh ./universe.pkl
> 90M	./universe.pkl

> 11M: trajectory
> 79M: topology
<MDAnalysis.core.topologyattrs.Atomnames object at 0x7fd41e2e2770>
5,4M	./dump_attr.pkl
<MDAnalysis.core.topologyattrs.AltLocs object at 0x7fd41e2e2b30>
5,4M	./dump_attr.pkl
<MDAnalysis.core.topologyattrs.ChainIDs object at 0x7fd41e2e2a10>
5,4M	./dump_attr.pkl
<MDAnalysis.core.topologyattrs.RecordTypes object at 0x7fd41e2e3520>
5,4M	./dump_attr.pkl
<MDAnalysis.core.topologyattrs.Atomids object at 0x7fd41e2e3160>
4,3M	./dump_attr.pkl
<MDAnalysis.core.topologyattrs.Tempfactors object at 0x7fd41e2e3e20>
4,3M	./dump_attr.pkl
<MDAnalysis.core.topologyattrs.Occupancies object at 0x7fd41b705c00>
4,3M	./dump_attr.pkl
<MDAnalysis.core.topologyattrs.Atomtypes object at 0x7fd48888f4f0>
5,4M	./dump_attr.pkl
<MDAnalysis.core.topologyattrs.Elements object at 0x7fd41b706440>
5,4M	./dump_attr.pkl
<MDAnalysis.core.topologyattrs.Masses object at 0x7fd41b705b40>
4,3M	./dump_attr.pkl
<MDAnalysis.core.topologyattrs.Resnums object at 0x7fd41b706410>
984K	./dump_attr.pkl
<MDAnalysis.core.topologyattrs.Resids object at 0x7fd41b705ba0>
984K	./dump_attr.pkl
<MDAnalysis.core.topologyattrs.Resnums object at 0x7fd41b705bd0>
984K	./dump_attr.pkl
<MDAnalysis.core.topologyattrs.ICodes object at 0x7fd41b7064d0>
1,3M	./dump_attr.pkl
<MDAnalysis.core.topologyattrs.Resnames object at 0x7fd41b705c30>
1,3M	./dump_attr.pkl
<MDAnalysis.core.topologyattrs.Segids object at 0x7fd41b706500>
4,0K	./dump_attr.pkl
<MDAnalysis.core.topologyattrs.Atomindices object at 0x7fd41b706770>
4,0K	./dump_attr.pkl
<MDAnalysis.core.topologyattrs.Resindices object at 0x7fd41b705ea0>
4,0K	./dump_attr.pkl
<MDAnalysis.core.topologyattrs.Segindices object at 0x7fd41b705e70>
4,0K	./dump_attr.pkl
```

**No need to redirect to the original frame during deserialization `self[self.ts.frame]` PR #3722**
- https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/coordinates/base.py#L1321
- for a 560k atom PDB single-frame trajectory, the serialization/deserialization time decreases from **2s** to **0.7s**.

```python
# with PR #3722
> dump time: 0.47 ± 0.03 s
> load time: 0.22 ± 0.01 s

90M	./universe.pkl
```

**lazily-building `Universe._topology.tt. _RA, _SR` and refactor `make_downshift_arrays` PR #3724**
- `Universe._topology.tt. _RA, _SR` are huge nested arrays that take a long time to serialize.
- for a 560k atom PDB topology, the serialization/deserialization time further decreases from **0.7s** to **0.2s**.

```python
# with PR #3724
> dump time: 0.11 ± 0.00 s
> load time: 1.37 ± 0.06 s

80M	./universe.pkl
```

**Decrease topology memory footprint**
- `nmidx` and `values` are pair-wise mapped. any reason to keep both inside the memory? (I know it's useful when one tries to modify the topology)

### LINKED ISSUE #3723: Fail to double-serialization for some coordinate Readers
## Expected behavior ##

All readers can be double serialized (i.e. serialize the newly deserialized Reader)

## Code to reproduce the behavior ##

``` python
import MDAnalysis as mda
import pickle
from MDAnalysis.tests.datafiles import XYZ

u = mda.Universe(XYZ)

u_new = pickle.loads(pickle.dumps(u))
u_new_dumped = pickle.dumps(u_new)

```
raises `AttributeError: 'TextIOPicklable' object has no attribute 'raw_class'`

## Current version of MDAnalysis ##

- Which version are you using? (run `python -c "import MDAnalysis as mda; print(mda.__version__)"`) dev
- Which version of Python (`python -V`)? 3.10
- Which operating system? Ubuntu 20.04

### LINKED ISSUE #4633: Coordinate Transformations Lost During Serialization
## Expected Behavior ##

Modifications to the coordinates, such as transformations, should be preserved during serialization.

## Actual Behavior ##

During serialization, the trajectory is reloaded based on the current frame, which removes any modifications to the coordinates, such as transformations.

## Code to reproduce the behavior ##

```python
import MDAnalysis as mda
from MDAnalysisTests.datafiles import TPR, XTC
from MDAnalysis.transformations.translate import translate
import pickle
u = mda.Universe(TPR, XTC)
u.trajectory[2]

print('initial x position:', u.atoms[0].position[0])
translate(vector=[1,0,0])(u.trajectory.ts)
print('x position after translation:', u.atoms[0].position[0])

u_pickled = pickle.loads(pickle.dumps(u))
print('x position after serialization:', u_pickled.atoms[0].position[0])
> initial x position: 54.560005
> x position after translation: 55.560005
> x position after serialization: 54.560005

# after the fix:
> x position after serialization: 55.560005

```

## Current version of MDAnalysis ##

- Which version are you using? (run `python -c "import MDAnalysis as mda; print(mda.__version__)"`)
dev
- Which version of Python (`python -V`)?
3.11
- Which operating system?
mac-os

### PR DESCRIPTION
**Pull request by yuxuanzhuang**
Fixes #3723
Fixes #4633 
Partial fix: #3721 

Changes made in this Pull Request:
 -  no redirection to the original frame (explicitly) during deserialization
 - TextIOPicklable serializes the raw class instance instead of class name.

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

### CODE REVIEW

**Comment by orbeckst** [comment]
Once you need someone to review, feel free to ping/request reviewers.

I'd also suggest using the "convert to draft" in GitHub instead of just adding "WIP" to the issue title as it shows more intentionally if you currently want reviews or not.

**Review by richardjgowers — CHANGES_REQUESTED** [review]
This is probably an improvement, but I think it'll be a change in behaviour so we'll need to raise warnings etc.

**Comment by yuxuanzhuang** [comment]
> This is probably an improvement, but I think it'll be a change in behaviour so we'll need to raise warnings etc.

The behaviour change is how the reading location `TextIOPicklable` is recovered. Before it is done by `universe.trajectory[i]`. Now it is directly serialized. Not sure if there's any explicitly warning that should be raised anywhere?

**Review by richardjgowers — COMMENTED** [review]
@yuxuanzhuang sorry for leaving this for a while.  So is the behaviour now that serialising/deserialising a Reader preserves the Timestep correctly?  I.e. if I'd modified the coordinates somehow, this data will be preserved?

**Comment by yuxuanzhuang** [comment]
@richardjgowers Okay, I never thought about preserving the modification of coordinates (before).

But the behavior does change after this PR, i.e. now the modification is preserved (because no data reloading from the file is performed after this [line](https://github.com/MDAnalysis/mdanalysis/pull/3722/files#diff-9d85ddb660f5430e254eb41665a572f70598bdd6d286130a52737d444e6d0fcaL1321) is removed) which is more "correct".

```python
u = mda.Universe(XYZ)

print('Orig pos:,', u.trajectory.ts.positions[0])
u.trajectory.ts.positions[0] = np.array([1, 2, 3])
print('New pos:,', u.trajectory.ts.positions[0])
u_new = pickle.loads(pickle.dumps(u))
print('After serializing pos:,', u_new.trajectory.ts.positions[0])

# original output
Orig pos:, [ 0.931 17.318 16.423]
New pos:, [1. 2. 3.]
After serializing pos:, [ 0.931 17.318 16.423]

# new output after #PR 3722
Orig pos:, [ 0.931 17.318 16.423]
New pos:, [1. 2. 3.]
After serializing pos:, [1. 2. 3.]
```

**Comment by richardjgowers** [comment]
@yuxuanzhuang thanks for adding the test - it looks like it found that h5md format has a custom serialisation mechanism

**Comment by yuxuanzhuang** [comment]
Should work now! (too many formats to keep track of :) )

**Comment by orbeckst** [comment]
@yuxuanzhuang could you resolve the conflicts please?

**Comment by orbeckst** [comment]
@yuxuanzhuang and @richardjgowers this PR still looks relevant. Why did it stall? What needs to be done to complete it?

**Comment by yuxuanzhuang** [comment]
> @yuxuanzhuang could you resolve the conflicts please?

Done! Should I worry about the linter bot?

**Comment by IAlibay** [comment]
> > @yuxuanzhuang could you resolve the conflicts please?
> 
> Done! Should I worry about the linter bot?

Unless it outlines a flake8 failure it's ok to ignore.

Just checked - no flake8 so you're good to go.

**Comment by yuxuanzhuang** [comment]
I understand why TNGReader needs to re-read the current frame during serialization; however, due to this, the modification of `ts` cannot be preserved in the TNG format. If I comment it out, all the tests will pass.
```
        # make sure we re-read the current frame to update C level objects in
        # the file iterator
        self._read_frame(self._frame)
```

Do you think it's reasonable to only make sure `_file_iterator` reads the step but no change of `ts`? @hmacdope
```
        # convert from frames to integrator steps
        step = self._frame_to_step(self._frame + 1)
        iterator_step = self._file_iterator.read_step(step)
````

**Comment by hmacdope** [comment]
> I understand why TNGReader needs to re-read the current frame during serialization; however, due to this, the modification of `ts` cannot be preserved in the TNG format. If I comment it out, all the tests will pass.
> 
> ```
>         # make sure we re-read the current frame to update C level objects in
>         # the file iterator
>         self._read_frame(self._frame)
> ```
> 
> Do you think it's reasonable to only make sure `_file_iterator` reads the step but no change of `ts`? @hmacdope
> 
> ```
>         # convert from frames to integrator steps
>         step = self._frame_to_step(self._frame + 1)
>         iterator_step = self._file_iterator.read_step(step)
> ```

Just checking my understanding @yuxuanzhuang, this means that the coordinate  data will be re-read, but that any *additional* data in `ts` will become invalid? I am just not sure exactly what is meant by `change of ts`. Perhaps a super quick example that demos the difference?

Sorry about this mess, tbh this is being caused by my lazyness from way back where I didn't make the cython level classes pickleable in `pytng` (its a bit complicated for `__cinnit__` using classes). See https://github.com/MDAnalysis/pytng/issues/97. If I am understanding correctly it would be better to just be able to pickle everything there, rather than forcing a re-read of the step? I can look at addressing this but may take a while as super busy. Perhaps better to go for an interim solution that you suggested above?

File: package/CHANGELOG [hunk 1]
```diff
@@ -48,6 +49,8 @@ Enhancements
     (Issue #3994, PR #4281)  
   * Add support for reading chainID info from Autodock PDBQT files (Issue #4207,
     PR #4284)
+  * Improve performance in Universe serialization by no explicit redirection to
```
Comment (hmacdope) [inline package/CHANGELOG]: Newest first :)

File: package/CHANGELOG [hunk 2]
```diff
@@ -198,6 +201,10 @@ Fixes
   * Fix MSD docs to use the correct error metric in example (Issue #3991)
   * Add 'PairIJ Coeffs' to the list of sections in LAMMPSParser.py
     (Issue #3336)
+  * Fix failure in double-serialization of TextIOPicklable file reader.
```
Comment (hmacdope) [inline package/CHANGELOG]: Same here.

File: package/CHANGELOG [hunk 3]
```diff
@@ -364,6 +371,8 @@ Changes
     use `pip install ./package[extra_formats]` instead (PR #3810)
   * `Universe.empty` emmits less warnings (PR #3814)
   * adding element attribute to TXYZParser if all atom names are valid element symbols (PR #3826)
+  * TextIOPicklable serializes the raw class instance instead of class name.
```
Comment (hmacdope) [inline package/CHANGELOG]: Same here.

**Review by hmacdope — CHANGES_REQUESTED** [review]
Tiny cosmetic things :)

**Comment by orbeckst** [comment]
@hmacdope and @richardjgowers  can you please check if your raised concerns have been addressed? 

If yes, can we merge?

Otherwise, are there bigger issues that we cannot solve?

File: package/MDAnalysis/coordinates/base.py [hunk 4]
```diff
@@ -659,6 +659,9 @@ class ProtoReader(IOBase, metaclass=_Readermeta):
     .. versionchanged:: 2.0.0
        Now supports (un)pickle. Upon unpickling,
        the current timestep is retained by reconstrunction.
+    .. versionchanged:: 2.7.0
```
Comment (orbeckst) [inline package/MDAnalysis/coordinates/base.py]:
```suggestion
    .. versionchanged:: 2.8.0
```

File: package/MDAnalysis/lib/picklable_file_io.py [hunk 5]
```diff
@@ -207,21 +203,34 @@ class TextIOPicklable(io.TextIOWrapper):
 
 
     .. versionadded:: 2.0.0
+    .. versionchanged:: 2.7.0
```
Comment (orbeckst) [inline package/MDAnalysis/lib/picklable_file_io.py]:
```suggestion
    .. versionchanged:: 2.8.0
```

**Review by orbeckst — COMMENTED** [review]
update versionchanged to 2.8.0

Note that I had to find the CHANGELOG entries from different sections to correctly merge. Please **double check the CHANGELOG** @yuxuanzhuang

**Comment by yuxuanzhuang** [comment]
> I am just not sure exactly what is meant by change of ts. Perhaps a super quick example that demos the difference?

I mean modification of e.g. coordinates.

```
u = mda.Universe(TNG_traj_gro, TNG_traj)
reader = u.trajectory
reader.ts.positions[0] = np.array([1, 2, 3])
reader_p = pickle.loads(pickle.dumps(reader))
assert_equal(reader.ts, reader_p.ts,
                       "Modification of ts not preserved after serialization")
```

Previously, during unpickling, `self._read_frame(self._frame)` will 1) update `self._file_iterator` to the current frame. 2) modify `self.ts` by reading from `self._file_iterator`. This will discard any change to the timestep itself.

Now, only `self._file_iterator` will be reopened and updated. `self.ts` will be preserved during serialization.

**Review by hmacdope — APPROVED** [review]
LGTM @yuxuanzhuang,  thanks for pushing through on this. :+1:

**Comment by orbeckst** [comment]
@richardjgowers is your review still pertinent or should I just dismiss it?

File: package/CHANGELOG [hunk 6]
```diff
@@ -22,6 +22,10 @@ The rules for this file:
  * 2.8.0
 
 Fixes
+ * Fix failure in double-serialization of TextIOPicklable file reader.
+   (Issue #3723, PR #3722)
+ * Fix failure to preserve modification of coordinates after serialization.
```
Comment (richardjgowers) [inline package/CHANGELOG]: There's a change in behaviour here where pickling then unpickling no longer resets a frame to the data held in that frame.  E.g. if I load frame 10, then unwrap these coordinates (modifying them from the file held state), then serialise/deserialise that (e.g. in a multiprocessing situation) the frame is now unwrapped on deserialisation.  This is probably a good thing (tm).  This should probably get explained in an issue and/or changelog entry

**Review by richardjgowers — CHANGES_REQUESTED** [review]

File: package/CHANGELOG (hunk 6 above)
Comment (yuxuanzhuang) [reply package/CHANGELOG]: Thanks! I have added an issue regarding it https://github.com/MDAnalysis/mdanalysis/issues/4633

**Review by richardjgowers — APPROVED** [review]

**Pull request merged**